File: lab/src/stats.ts
import type { FontSignalsPayload } from "@fontmask/collector";
import type { MaskingConfig, MaskingPresetId } from "@fontmask/config";
import { sanitizeMaskingConfig } from "@fontmask/config";
import crypto from "node:crypto";

/**
 * Describes one cohort fingerprint row annotated with hashing metadata plus engine labels.
 */
export type CohortRow = {
  id: number;
  label: string;
  payload: FontSignalsPayload;
  hashHex: string;
  enginePreset: MaskingPresetId | "unset";
};

/**
 * Flattens font metrics into numeric vectors powering cosine correlation estimates.
 */
export const vectorizePayload = (payload: FontSignalsPayload): number[] => {
  const widthValues = sortedKeys(payload.measureTextWidths).map(
    (key) => payload.measureTextWidths[key] ?? 0
  );
  const phantomValues = sortedKeys(payload.phantomProbeMetrics).map(
    (key) => payload.phantomProbeMetrics[key] ?? 0
  );
  return [...widthValues, ...phantomValues, payload.fontsCheckHits.length];
};

const sortedKeys = (record: Record<string, number>): string[] =>
  Object.keys(record).sort();

/**
 * Computes normalized cosine similarity treating missing dimensions as orthogonal noise.
 */
export const cosineSimilarity = (left: number[], right: number[]): number => {
  const length = Math.max(left.length, right.length);
  let dot = 0;
  let leftNorm = 0;
  let rightNorm = 0;
  for (let idx = 0; idx < length; idx += 1) {
    const l = left[idx] ?? 0;
    const r = right[idx] ?? 0;
    dot += l * r;
    leftNorm += l * l;
    rightNorm += r * r;
  }
  if (leftNorm === 0 || rightNorm === 0) {
    return 0;
  }
  return dot / (Math.sqrt(leftNorm) * Math.sqrt(rightNorm));
};

/**
 * Estimates pairwise cosine correlation statistics for uniqueness comparisons.
 */
export const summarizePairwiseCosine = (rows: CohortRow[]) => {
  const vectors = rows.map((row) => vectorizePayload(row.payload));
  const scores: number[] = [];
  for (let outer = 0; outer < rows.length; outer += 1) {
    for (
      let inner = outer + 1;
      inner < rows.length;
      inner += 1
    ) {
      scores.push(cosineSimilarity(vectors[outer]!, vectors[inner]!));
    }
  }
  scores.sort((a, b) => a - b);
  const mean =
    scores.length === 0
      ? 0
      : scores.reduce((acc, score) => acc + score, 0) / scores.length;
  const p95 =
    scores.length === 0
      ? 0
      : scores[Math.floor(0.95 * (scores.length - 1))] ?? scores.at(-1) ?? 0;
  const min = scores[0] ?? 0;
  const max = scores.at(-1) ?? 0;
  return {
    scores,
    mean,
    min,
    max,
    p95,
    countPairs: scores.length,
  };
};

/**
 * Calculates collision counts referencing stable SHA digests of payloads.
 */
export const analyzeCollisions = (rows: CohortRow[]) => {
  const tally = new Map<string, number>();
  rows.forEach((row) => {
    tally.set(row.hashHex, (tally.get(row.hashHex) ?? 0) + 1);
  });
  const collisions = [...tally.values()].filter((count) => count > 1).length;
  return {
    uniqueHashes: tally.size,
    collisions,
  };
};

/**
 * Estimates Shannon entropy on discretized metric streams per cohort column.
 */
export const estimateMetricEntropy = (rows: CohortRow[]): number => {
  if (rows.length === 0) {
    return 0;
  }
  const buckets = new Map<string, number>();
  rows.forEach((row) => {
    const vector = vectorizePayload(row.payload);
    const label = vector.map((value) => value.toFixed(4)).join("|");
    buckets.set(label, (buckets.get(label) ?? 0) + 1);
  });
  const total = rows.length;
  let entropy = 0;
  buckets.forEach((count) => {
    const p = count / total;
    entropy -= p * Math.log2(p);
  });
  return entropy;
};

/**
 * Serializes cohort artifacts for downstream visualization layers.
 */
export const serializeCohortSummary = (rows: CohortRow[]) => ({
  rows: rows.length,
  collisions: analyzeCollisions(rows),
  pairwiseCosine: summarizePairwiseCosine(rows),
  entropy: estimateMetricEntropy(rows),
});

/**
 * Serializes payloads into stable SHA-256 digests for collision tracking.
 */
export const hashPayload = (payload: FontSignalsPayload): string => {
  const stable = canonicalizePayload(payload);
  return crypto.createHash("sha256").update(stable).digest("hex");
};

const canonicalizePayload = (payload: FontSignalsPayload): string =>
  JSON.stringify(payload, Object.keys(payload).sort());

/**
 * Mirrors runtime masking clamps for lab-only synthetic experiments.
 */
export const labMask = (masking: MaskingConfig): MaskingConfig =>
  sanitizeMaskingConfig(masking);

File: python/fontmask_lab/stats.py
from __future__ import annotations

import math
from typing import Any


def vectorize_payload(payload: dict[str, Any]) -> list[float]:
    widths = sorted(payload.get("measureTextWidths") or {})
    phantom = sorted(payload.get("phantomProbeMetrics") or {})
    width_vals = [float((payload.get("measureTextWidths") or {}).get(k, 0)) for k in widths]
    phantom_vals = [float((payload.get("phantomProbeMetrics") or {}).get(k, 0)) for k in phantom]
    hits = payload.get("fontsCheckHits") or []
    return [*width_vals, *phantom_vals, float(len(hits))]


def cosine_similarity(left: list[float], right: list[float]) -> float:
    length = max(len(left), len(right))
    dot = 0.0
    ln = 0.0
    rn = 0.0
    for idx in range(length):
        l = left[idx] if idx < len(left) else 0.0
        r = right[idx] if idx < len(right) else 0.0
        dot += l * r
        ln += l * l
        rn += r * r
    if ln == 0 or rn == 0:
        return 0.0
    return dot / (math.sqrt(ln) * math.sqrt(rn))


def summarize_pairwise_cosine(rows: list[dict[str, Any]]) -> dict[str, Any]:
    vectors = [vectorize_payload(r["payload"]) for r in rows]
    scores: list[float] = []
    for outer in range(len(rows)):
        for inner in range(outer + 1, len(rows)):
            scores.append(cosine_similarity(vectors[outer], vectors[inner]))
    scores.sort()
    n = len(scores)
    mean = 0.0 if n == 0 else sum(scores) / n
    p95 = (
        0.0
        if n == 0
        else scores[int(0.95 * (n - 1))]
    )
    return {
        "scores": scores,
        "mean": mean,
        "min": scores[0] if scores else 0.0,
        "max": scores[-1] if scores else 0.0,
        "p95": p95,
        "countPairs": n,
    }


def analyze_collisions(rows: list[dict[str, Any]]) -> dict[str, int]:
    tally: dict[str, int] = {}
    for row in rows:
        h = row["hashHex"]
        tally[h] = tally.get(h, 0) + 1
    collisions = sum(1 for c in tally.values() if c > 1)
    return {"uniqueHashes": len(tally), "collisions": collisions}


def estimate_metric_entropy(rows: list[dict[str, Any]]) -> float:
    if not rows:
        return 0.0
    buckets: dict[str, int] = {}
    for row in rows:
        vector = vectorize_payload(row["payload"])
        label = "|".join(f"{v:.4f}" for v in vector)
        buckets[label] = buckets.get(label, 0) + 1
    total = len(rows)
    entropy = 0.0
    for count in buckets.values():
        p = count / total
        entropy -= p * math.log2(p)
    return entropy


def build_correlation_matrix(rows: list[dict[str, Any]]) -> list[list[float]]:
    vectors = [vectorize_payload(r["payload"]) for r in rows]
    size = min(len(vectors), 12)
    grid: list[list[float]] = []
    for outer in range(size):
        line: list[float] = []
        for inner in range(size):
            line.append(cosine_similarity(vectors[outer], vectors[inner]))
        grid.append(line)
    return grid
